src/app/courses/courses-list/courses-list.component.ts

```diff
@@ -1,62 +1,102 @@
-import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
+import { Component, OnInit, Input, OnChanges, OnDestroy } from '@angular/core';
+import { debounceTime } from 'rxjs';
 
-import { Course } from 'src/app/shared/interfaces/course.interface.';
-import { FilterPipe } from 'src/app/shared/pipes/filter.pipe';
+import { HttpCourse } from 'src/app/shared/interfaces/http-course.interface';
 import { CourseService } from 'src/app/shared/services/course.service';
 
 @Component({
   selector: 'app-courses-list',
   templateUrl: './courses-list.component.html',
   styleUrls: ['./courses-list.component.css']
 })
-export class CoursesListComponent implements OnInit {
-
-  constructor(private coursesService: CourseService) {}
+export class CoursesListComponent implements OnInit, OnChanges {
 
   @Input() searchValue: string = '';
-
-  courses: Course[] = [];
-  courseForDelete?: Course;
-  titleForDelete = "";
+  courseForDelete?: HttpCourse;
+  titleForDelete: string = '';
   showModal = false;
+  listOfCourses: HttpCourse[] = [];
+  inputValue = '';
+  debounceTimeValue = 500;
 
-
-  ngOnInit() {
-      this.courses = this.coursesService.getListCourses();
+  constructor(private courseService: CourseService ) {
+    
   }
-
-  trackByFn(index: number, item: Course): number {
-    return item.id; // Assuming each item has a unique "id" property
+  
+  ngOnInit() {
+    this.courseService.loadingSubject.next(true);
+    this.courseService.onFetchCourses()
+    .subscribe(
+      courses => {
+        this.listOfCourses = [...courses];
+      }
+    )
+    this.courseService.loadingSubject.next(false);
   }
-
-  filterItems(courses: Course[], searchtext: string): Course[] {
-    let filterPipe = new FilterPipe();
-    return filterPipe.transform(courses, searchtext);
+    
+  ngOnChanges() {
+    this.courseService.inputSubject
+    .pipe(debounceTime(this.debounceTimeValue))
+    .subscribe(value => {
+      this.inputValue = value;
+      if (this.inputValue.length >= 3) {
+        this.courseService.loadingSubject.next(true);
+        this.courseService.onFetchCoursesWithFilter(this.inputValue).subscribe(
+          courses => {
+            this.listOfCourses = [...courses];
+          }
+        )
+        this.courseService.loadingSubject.next(false);
+      } else {
+        this.courseService.loadingSubject.next(true);
+        this.courseService.onFetchCourses().subscribe(
+          courses => {
+            this.listOfCourses = [...courses];
+          }
+        )
+        this.courseService.loadingSubject.next(false);
```

Comment:
I see why you use this block - but the filteting has to be in the pipe
Please change the logic of the courses uploading

Reply:
Did I understand correctly that I should use new customPipe in template for filtering like `*ngIf="listOfCourses | filter "`?
If yes - I tried, but my pipe logic uses subscription and this pipe returns initial value without waiting the response of subscription.
So filter is not rendering in a proper way. I tried to use async custom pipe with returning of promise, but it still doesn't render the filter results.

Reply:
I meant that it is better to use the **filter operator** with the help of **RxJs**
Pipe is not needed here, filtering for the data should be done in the scope of the RxJs operators